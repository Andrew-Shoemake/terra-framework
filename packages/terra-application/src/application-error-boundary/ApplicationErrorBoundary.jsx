import React from 'react';
import PropTypes from 'prop-types';
import StatusView from 'terra-status-view';

const propTypes = {
  /**
   * Components to render within the context of the ApplicationErrorBoundary
   */
  children: PropTypes.node,
};

class ApplicationErrorBoundary extends React.Component {
  constructor(props) {
    super(props);

    this.resetError = () => {
      this.errorRef.current = undefined;
      if (this.state.error) {
        this.setState({ error: undefined });
      }  
    }

    this.errorRef = React.createRef();

    this.state = { error: undefined };
  }

  static getDerivedStateFromError(error) {
    return { error };
  }

  componentDidCatch(error) {
    this.errorRef.current = error;
    this.setState({ error: false });
  }

  componentDidMount() {
    this.resetError();
  }

  componentDidUpdate() {
    this.resetError();
  }

  render() {
    const { children } = this.props;
    const activeError = this.state.error || this.errorRef.current;

    if (activeError) {
      return (
        <StatusView
          variant="error"
          message={activeError.toString()}
        />
      );
    }

    return children;
  }
}

ApplicationErrorBoundary.propTypes = propTypes;

export default ApplicationErrorBoundary;
