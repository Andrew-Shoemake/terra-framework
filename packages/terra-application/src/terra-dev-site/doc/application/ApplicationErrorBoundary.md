# Terra Application

## Application Error Boundary

The ApplicationErrorBoundary allows for the handling of render-time errors within the Terra Application framework.

> Note: ApplicationBase renders an ApplicationErrorBoundary by default, but additional ApplicationErrorBoundary components
can be used to scope errors within different areas of an application.
