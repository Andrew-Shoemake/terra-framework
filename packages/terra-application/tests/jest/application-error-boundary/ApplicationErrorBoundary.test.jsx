import React from 'react';
import { mockIntl, shallowWithIntl } from 'terra-enzyme-intl';
import ApplicationErrorBoundary from '../../../src/application-error-boundary/ApplicationErrorBoundary';

describe('ApplicationErrorBoundary', () => {
  describe('Snapshots', () => {
    it('should render with minimal props', () => {
      const wrapper = shallowWithIntl((
        <ApplicationErrorBoundary />
      ));

      expect(wrapper).toMatchSnapshot();
    });

    it('should render with children', () => {
      const wrapper = shallowWithIntl((
        <ApplicationErrorBoundary>
          <div>Test child</div>
        </ApplicationErrorBoundary>
      ));

      expect(wrapper).toMatchSnapshot();
    });

    it('should render error view when an error is detected', () => {
      const wrapper = shallow((
        <ApplicationErrorBoundary.WrappedComponent
          intl={mockIntl}
        >
          <div>Test child</div>
        </ApplicationErrorBoundary.WrappedComponent>
      ));

      wrapper.setState({ hasError: true });

      expect(wrapper).toMatchSnapshot();
    });
  });
});
