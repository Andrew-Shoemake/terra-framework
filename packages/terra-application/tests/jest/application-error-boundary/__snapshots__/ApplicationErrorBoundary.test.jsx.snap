// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationErrorBoundary Snapshots should render error view when an error is detected 1`] = `
<StatusView
  buttonAttrs={Array []}
  isAlignedTop={false}
  isGlyphHidden={false}
  message="Error: test error"
  variant="error"
/>
`;

exports[`ApplicationErrorBoundary Snapshots should render error view when an error is detected 2`] = `<ErrorComponent />`;

exports[`ApplicationErrorBoundary Snapshots should render with children 1`] = `
<div>
  Test child
</div>
`;

exports[`ApplicationErrorBoundary Snapshots should render with minimal props 1`] = `""`;
